### PR TITLE
Add geometry vectors module and initial `Piece` logic

### PR DESCRIPTION
**Pull request by kpiotr6**
Adding vectors and their tests.

### CODE REVIEW

**Comment by Szaroslav** [comment]
Prettified the code; added semicolons and multi-line imports/exports of bigger objects :)

You need to change imports/exports from `.js` to `.ts` extensions, we may to update `tsconfig.json` or `package.json` (don't remember) to avoid including these extensions constantly. Also, add 120 width rule in _VSCode_ and adjust vector test module to it.
1. CTRL + Shift + P (open command palette).
2. Type **Open user settings** and open JSON file.
3. Add new property:
```json
"editor.rulers": [ 120 ]
```
<br>

You accidentally remove at least 2 lines from `package.json` - `dev` and `build` scripts.

**Review by Szaroslav — CHANGES_REQUESTED** [review]
Prettified the code; added semicolons and multi-line imports/exports of bigger objects :)

You need to change imports/exports from `.js` to `.ts` extensions, we may to update `tsconfig.json` or `package.json` (don't remember) to avoid including these extensions constantly. Also, add 120 width rule in _VSCode_ and adjust vector test module to it.
1. CTRL + Shift + P (open command palette).
2. Type **Open user settings** and open JSON file.
3. Add new property:
```json
"editor.rulers": [ 120 ]
```
<br>

You accidentally remove at least 2 lines from `package.json` - `dev` and `build` scripts.

**Comment by Szaroslav** [comment]
I will check the geometry module and its tests later, but looks good.

File: tests/utils/vector.test.ts
Comment (Szaroslav) [inline tests/utils/vector.test.ts:1]: Wrong directory.

File: src/geometry/vector2d.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    this.y = y;
+  }
+
+  public getX() {
+    return this.x;
+  }
+
+  public getY() {
+    return this.y;
+  }
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:29]: For new we don't need getters, also TypeScript offers explicit getters and setters: [TypeScript getters/setters](https://www.typescriptlang.org/docs/handbook/2/classes.html#getters--setters).

File: src/geometry/vector2d.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+
+export class Vector2d {
+  protected readonly x: number;
+  protected readonly y: number;
+
+  protected createVector(x: number, y: number): Vector2d {
+    return new Vector2d(x, y);
+  }
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:16]: I would swap this method position with `constructor`.

File: src/geometry/vector2d.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+export class BoardVector2d extends IntVector2d {
+
+  protected createVector(x: int, y: int): Vector2d {
+    return new BoardVector2d(x, y);
+  }
+
+  public static fromVector2d(vector: Vector2d | IntVector2d): BoardVector2d {
+    let board: BoardVector2d = vector as BoardVector2d
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:124]: Wrong variable name, it's very confusing.

File: src/geometry/vector2d.ts [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      case Symmetry.X_AXIS:
+        return this.createVector(this.x, -this.y);
+
+      case Symmetry.Y_AXIS:
+        return this.createVector(-this.x, this.y);
+    }
+  }
+}
+
+export class IntVector2d extends Vector2d {
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:95]: Need to hide `int` implementation due continuous type casting.

File: tests/geometry/vector.test.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  test("Equality", () => {
+    expect(new Vector2d(1.2, 1)).toStrictEqual(new Vector2d(1.2, 1));
+    expect(new Vector2d(1, 10)).not.toStrictEqual(new Vector2d(1, 1));
+    expect(new Vector2d(1, 1)).not.toStrictEqual(new BoardVector2d(1 as int, 1 as int));
+  });
+
+  test("Multiplying/dividing 2 vectors", () => {
+    expect(new Vector2d(1, 2).mul(new Vector2d(2, 1.5))).toStrictEqual(new Vector2d(2, 3));
+    expect(new Vector2d(1, 2).div(new Vector2d(2, 2))).toStrictEqual(new Vector2d(0.5, 1));
+    expect(()=>{
```
Comment (Szaroslav) [inline tests/geometry/vector.test.ts:25]: Spaces :) `expect(() => {`

File: src/geometry/vector2d.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      return this.createVector(this.x * other.x, this.y * other.y);
+    }
+    return this.createVector(this.x * other, this.y * other);
+  }
+
+  public div(other: Vector2d | number): Vector2d {
+    if (other instanceof Vector2d) {
+      if(other.x == 0 || other.y == 0){
+        throw new Error("At least one coordinate of divider vector is equal to 0.");
+      }
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:61]: Should split the condition into separate dimensions for the clarity of exception. Also, add space after `if` keyword and closing bracket.

File: src/geometry/vector2d.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      return this.createVector(this.x / other.x, this.y / other.y);
+    }
+    if(other == 0){
```
Comment (Szaroslav) [inline src/geometry/vector2d.ts:64]: Add space after `if` keyword and closing bracket.

File: tests/geometry/vector.test.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+    expect(new BoardVector2d(1 as int, 3 as int).yToString()).toStrictEqual("4");
+    expect(new BoardVector2d(1 as int, 3 as int).toString()).toStrictEqual("b4");
+  });
+
+
+  test("Copy vector", () => {
+    expect(new BoardVector2d(1 as int, 1 as int).copy()).toStrictEqual(new BoardVector2d(1 as int, 1 as int));
+    expect(new BoardVector2d(1 as int, 2 as int).copy()).toStrictEqual(new BoardVector2d(1 as int, 2 as int));
+  });
+});
```
Comment (Szaroslav) [inline tests/geometry/vector.test.ts:107]: Should test also edge cases, when these vectors aren't constructed by integer numbers.

**Review by Szaroslav — CHANGES_REQUESTED** [review]

**Review by Szaroslav — CHANGES_REQUESTED** [review]
Keep going, keep grinding. All eyes on you.

File: src/geometry/vector2d.ts [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    if (other instanceof Vector2d) {
+      if (other.x == 0) {
+        throw new Error("First (x) coordinate of divider vector is equal to 0.");
+      }
+      if (other.y == 0) {
+        throw new Error("Second (y) coordinate of divider vector is equal to 0.");
+      }
+      return this.createVector(this.x / other.x, this.y / other.y);
+    }
+    if (other == 0) {
```
[the remark this answers is not in the document]
Comment (kpiotr6) [reply src/geometry/vector2d.ts:147]: Yeah. Don't remind me xd.

File: src/pieces/direction.ts
Comment (Szaroslav) [inline src/pieces/direction.ts:1]: Understood xd

**Review by Szaroslav — APPROVED** [review]
Some changes need to be applied but I closing this pull request. Need to open a new one with part of the game logic.
